torch 4 (final in series) (#113)
rstudio/ai-blog: _posts/2020-10-09-torch-optim/torch_network_with_optim.Rmd

---
title: "Optimizers in torch"
description: > 

>  Today, we wrap up our mini-series on torch basics, adding to our toolset two abstractions we hadn't made use of before: loss functions and optimizers.

Reviewer: ```suggestion
 Today, we wrap up our mini-series on torch basics, adding to our toolset two abstractions: loss functions and optimizers.
```

Reviewer: That they weren't used before is implied by the fact that you're adding them to the toolset 👍

Author: lol right

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/

> slug: keydanatorchnetworkfromscratch

Reviewer: Is this the right slug for this post?

Author: this too...

> date: 10-01-2020

Reviewer: If I'm understanding things correctly, I think this is supposed to be the ninth of October.

Author: thanks for spotting this!

categories:
  - Torch
  - R
bibliography: bibliography.bib
output:
  distill::distill_article:
    self_contained: false
    toc: true
preview: images/preview.jpg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

```{r}

```

> This is the fourth and last installment in a series introducing `torch` basics. Initially, we [focused on *tensors*](https://blogs.rstudio.com/ai/posts/2020-10-01-torch-network-from-scratch/). To illustrate their power, we coded a complete (if toy-size) neural network from scratch. We didn't make use of any of `torch`'s higher-level capabilities -- not even *autograd*, its automatic-differentiation feature.

Reviewer: I would rework this to put the autograd description with the second post (e.g., In the follow-up post, we made use of autograd, torch's automatic differentiation feature…).

This changed in the [follow-up post](https://blogs.rstudio.com/ai/posts/2020-10-05-torch-network-with-autograd). No more thinking about derivatives and the chain rule; a single call to `backward()` did it all.

> [Two days ago](https://blogs.rstudio.com/ai/posts/2020-10-07-torch-modules), the code again saw a major simplification. Modules, and compositions thereof, turned out to be much more manageable than tensors, hand-assembled in some DAG[^1].

Reviewer: ```suggestion
[In the third post](https://blogs.rstudio.com/ai/posts/2020-10-07-torch-modules), the code again saw a major simplification. Modules, and compositions thereof, turned out to be much more manageable than tensors, hand-assembled in some DAG[^1].
```

Reviewer: That last sentence feels a little clunky right now—I think you can make the language quite simple (e.g. we used modules to…), I'm just not sure what it should be.

Author: yeah, right... tried to rephrase it some...

[^1]: directed acyclic graph

> Based on that last state, there are just two more things to do. For one, we still compute the loss by hand. And secondly, even though we get the gradients all nicely computed from *autograd*, we still loop over the model's parameters, updating them all ourselves. You won't be surprised to hear that nothing of this is necessary.

Reviewer: > Based on that last state,

Does this mean you're picking up in the code where you left off in the third post? If not, that's kind of what it sounds like (since you're referencing "state").

Author: yeah, that's the intention - couldn't think of how to say this better :-(

> Based on that last state, there are just two more things to do. For one, we still compute the loss by hand. And secondly, even though we get the gradients all nicely computed from *autograd*, we still loop over the model's parameters, updating them all ourselves. You won't be surprised to hear that nothing of this is necessary.

Reviewer: ```suggestion
Based on that last state, there are just two more things to do. For one, we still compute the loss by hand. And secondly, even though we get the gradients all nicely computed from *autograd*, we still loop over the model's parameters, updating them all ourselves. You won't be surprised to hear that none of this is necessary.
```

Reviewer: That last sentence is like a triple negative, and I'm not sure where it leaves us… Maybe cut it? My correction is there in case you keep it.

## Losses and loss functions

> `torch` comes with all the usual loss functions, such as mean squared error, cross entropy, Kullback-Leibler divergence, and their likes. In general, there are two usage modes.

Reviewer: ```suggestion
`torch` comes with all the usual loss functions, such as mean squared error, cross entropy, Kullback-Leibler divergence, and the like. In general, there are two usage modes.
```

Reviewer: "and the like" is a little idiom, so it doesn't get pluralized (or whatever) like literal language would be.

> Taking the example of mean squared error, one way is to call `nnf_mse_loss()` directly on the prediction and ground truth tensors. For example:

Reviewer: ```suggestion
Taking the example of calculating mean squared error, one way is to call `nnf_mse_loss()` directly on the prediction and ground truth tensors. For example:
```

Reviewer: Or you could move calculating to the second part of the sentence (one way to calculate this is to call…) Something other than "calculating" would work too, you just need an action to go with mean squared error.

Author: yeah... even split up the sentence in two now :-)

```{r}
x <- torch_randn(c(3, 2, 3))
y <- torch_zeros(c(3, 2, 3))

nnf_mse_loss(x, y)
```

    torch_tensor 
    0.682362
    [ CPUFloatType{} ]

Other loss functions designed to be called directly start with `nnf_` as well: `nnf_binary_cross_entropy()`, `nnf_nll_loss()`, `nnf_kl_div()` ... and so on. [^2]

[^2]: The prefix `nnf_` was chosen because in PyTorch, the corresponding functions live in [torch.nn.functional](https://pytorch.org/docs/stable/nn.functional.html).

The second way is to define the algorithm in advance and call it at some later time. Here, respective constructors all start with `nn_` and end in `_loss`. For example: `nn_bce_loss()`, `nn_nll_loss(),` `nn_kl_div_loss()` ... [^3]

[^3]: This time, the corresponding PyTorch module is [torch.nn](https://pytorch.org/docs/stable/nn.html).

```{r}
loss <- nn_mse_loss()

loss(x, y)
```

    torch_tensor 
    0.682362
    [ CPUFloatType{} ]

> This method may be preferable when one and the same algorithm, as specified by configurable parameters, should be applied to more than one pair of tensors.

Reviewer: ```suggestion
This method may be preferable when the same algorithm should be applied to more than one pair of tensors, as specified by configurable parameters.
```

Author: this one is tricky... if if move that phrase it doesn't quite convey the intention ... clear sign the whole thing is too imprecise

      Public:
        add_param_group: function (param_group) 
        clone: function (deep = FALSE) 
        defaults: list
        initialize: function (params, lr = 0.001, betas = c(0.9, 0.999), eps = 1e-08, 
        param_groups: list
        state: list
        step: function (closure = NULL) 
        zero_grad: function () 

At any time, we can inspect those parameters:

```{r}
optimizer$param_groups[[1]]$params
```

    $weight
    torch_tensor 
    -0.0385  0.1412 -0.5436
    [ CPUFloatType{1,3} ]

    $bias
    torch_tensor 
    -0.1950
    [ CPUFloatType{1} ]

Now we perform the forward and backward passes. The backward pass calculates the gradients, but does *not* update the parameters, as we can see both from the model *and* the optimizer objects:

```{r}
out <- model(data)
out$backward()

optimizer$param_groups[[1]]$params
model$parameters
```

    $weight
    torch_tensor 
    -0.0385  0.1412 -0.5436
    [ CPUFloatType{1,3} ]

    $bias
    torch_tensor 
    -0.1950
    [ CPUFloatType{1} ]

    $weight
    torch_tensor 
    -0.0385  0.1412 -0.5436
    [ CPUFloatType{1,3} ]

    $bias
    torch_tensor 
    -0.1950
    [ CPUFloatType{1} ]

> Calling `step()` on the optimizer actually *performs* the updates. Again, let's be paranoid and check that both model and optimizer now hold the updated values:

Reviewer: ```suggestion
Calling `step()` on the optimizer actually *performs* the updates. Again, let's check that both model and optimizer now hold the updated values:
```
